cleverhans/dataset.py

```diff
@@ -0,0 +1,146 @@
+"""Dataset class for CleverHans
+
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from cleverhans.utils_mnist import data_mnist
+
+keras = None  # Only load keras if user tries to use a dataset that requires it
+
+
+class Dataset(object):
+    """Abstract base class representing a dataset.
+    """
+
+    # The number of classes in the dataset. Should be specified by subclasses.
+    nb_classes = None
+
+    def get_factory(self):
+        """Returns a picklable callable that recreates the dataset.
+        """
+
+        if hasattr(self, 'args'):
+            args = self.args
+        else:
+            args = []
+
+        if hasattr(self, 'kwargs'):
```

Comment:
No need to test, simply make a class constructor such that kwargs are always required.

```
def __init__(self, kwargs):
  if 'self' in kwargs:
   del kwargs['self']
  self.kwargs = kwargs
```

Reply:
Mostly I was thinking that some classes might not take any arguments at all. For these classes, it's slightly more convenient if the child class doesn't need to set self.kwargs = {} or call super(Subclass, self).__init__ to make Dataset.__init__ set self.kwargs = {}.